[RFC] ASoC: Add Multi CPU DAI support

bardliao opened the pull request:
This is a rebase version of https://www.alsa-project.org/pipermail/alsa-devel/2018-June/137190.html

The original commit message is:
```
As discussed in [1], ASoC core supports multi codec DAIs
on a DAI link. However it does not do so for CPU DAIs.

So, add support for multi CPU DAIs on a DAI Link by adding
multi CPU DAI in Card instantiation, suspend and resume
functions, PCM ops, stream handling functions and DAPM.

[1]: https://www.spinics.net/lists/alsa-devel/msg71369.html
```

Thanks @ranj063 for the comments. I addressed some, but I need more
time or help from others for the other comments.

ranj063 commented on sound/soc/soc-core.c:
```diff
@@ -2212,11 +2237,13 @@ int snd_soc_poweroff(struct device *dev)
 		int i;
 
-		pinctrl_pm_select_sleep_state(cpu_dai->dev);
+		for_each_rtd_codec_dai(rtd, i, cpu_dai) {
```
@bardliao typo here. For_each_cpu_dai?

bardliao replied:
Thank will update it.

ranj063 commented on sound/soc/soc-core.c:
```diff
@@ -785,10 +802,10 @@ int snd_soc_resume(struct device *dev)
-
-		if (cpu_dai->active)
-			pinctrl_pm_select_default_state(cpu_dai->dev);
+		for_each_rtd_codec_dai(rtd, i, cpu_dai) {
```
Looks like typo here as well? For each cpu dai?

bardliao replied:
Indeed. Thanks.

ranj063 commented on sound/soc/soc-core.c:
```diff
@@ -491,6 +491,14 @@ static struct snd_soc_pcm_runtime *soc_new_pcm_runtime(
 
+	/*
+	 * for rtd->cpu_dais
+	 */
```
This comment is not very helpful at all

bardliao replied:
This is to be consistent with the existing code. ie
```
	/*
	 * for rtd->codec_dais
	 */
```
above.

ranj063 commented on sound/soc/soc-pcm.c:
```diff
@@ -1430,10 +1546,18 @@ static int dpcm_prune_paths(struct snd_soc_pcm_runtime *fe, int stream,
-		/* prune the BE if it's no longer in our active list */
-		if (widget && widget_in_list(list, widget))
+			/*
+			 * The BE is pruned only if none of the the cpu_dai
```
Duplicate "the" in comment

bardliao replied:
Thank will update it.

plbossart commented on include/sound/soc.h:
```diff
@@ -1149,6 +1154,9 @@ struct snd_soc_pcm_runtime {
 
+	struct snd_soc_dai **cpu_dais;
+	unsigned int num_cpu_dai;
+
```
I wonder why we need to keep 'codec_dai' now that we have 'codec_dais" and 'num_codecs'?

Also why do we need 'cpu_dai' as well as 'cpus_dais' and 'num_cpu_dai'

Looks like room for additional cleanups

bardliao replied:
> I wonder why we need to keep 'codec_dai' now that we have 'codec_dais" and 'num_codecs'?
> 
> Also why do we need 'cpu_dai' as well as 'cpus_dais' and 'num_cpu_dai'
> 
> Looks like room for additional cleanups

Indeed, we can send another patch to remove cpu_dai and codec_dai

plbossart commented on sound/soc/soc-core.c:
```diff
@@ -1658,33 +1679,35 @@ int snd_soc_runtime_set_dai_fmt(struct snd_soc_pcm_runtime *rtd,
-	if (cpu_dai->component->driver->non_legacy_dai_naming) {
-		unsigned int inv_dai_fmt;
+	for_each_rtd_cpu_dai(rtd, i, cpu_dai) {
+		if (cpu_dai->component->driver->non_legacy_dai_naming) {
```
what is the 'legacy_dai_naming' and can it be removed now?

bardliao replied:
> what is the 'legacy_dai_naming' and can it be removed now?

@plbossart No, I don't think it can be removed now. Codec will be registered to cpu dai in CODEC <-> CODEC link, and we can use the non_legacy_dai_naming flag to know that is codec.

plbossart commented on sound/soc/soc-pcm.c:
```diff
@@ -920,17 +1003,19 @@ static int soc_pcm_hw_params(struct snd_pcm_substream *substream,
+	for_each_rtd_cpu_dai(rtd, i, cpu_dai) {
+		ret = snd_soc_dai_hw_params(cpu_dai, substream, params);
+		if (ret < 0)
+			goto interface_err;
```
you need to add the error handling here, the interface_err only deals with the codec case

bardliao replied:
> you need to add the error handling here, the interface_err only deals with the codec case

Yes, thanks @plbossart

bardliao replied:
@plbossart Updated

plbossart commented on sound/soc/soc-pcm.c:
```diff
@@ -1430,10 +1546,18 @@ static int dpcm_prune_paths(struct snd_soc_pcm_runtime *fe, int stream,
+			if (widget && widget_in_list(list, widget))
+				do_prune = 0;
+		}
+		if (!do_prune)
```
this will conflict with @lyakh 's work...

bardliao replied:
> this will conflict with @lyakh 's work...

Sorry, @plbossart @lyakh Could you describe it in more detail? Which PR will be affected?

lyakh replied:
@bardliao it's #1501 but I think it should be ok for you to proceed with your PR for now. I'm not sure what @plbossart's plan is, but as far as I understand, even after #1501 is fully reviewed and approved here, we'll first need to also get it reviewed on ALSA and virtualisation lists, so, it will take a while.

plbossart commented on sound/soc/soc-pcm.c:
```diff
@@ -3038,7 +3172,7 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 	dev_info(rtd->card->dev, "%s <-> %s mapping ok\n",
 		 (rtd->num_codecs > 1) ? "multicodec" : rtd->codec_dai->name,
-		 cpu_dai->name);
+		 (rtd->num_cpu_dai > 1) ? "multicpu" : rtd->cpu_dai->name);
```
this looks really ugly, what cant' we use rtd->cpu_dais[0].name and completely remove the use of rtd->cpu_dai?

plbossart commented on sound/soc/soc-dapm.c:
```diff
@@ -4295,63 +4294,74 @@ static void dapm_connect_dai_link_widgets(struct snd_soc_card *card,
+	for_each_rtd_codec_dai(rtd, i, codec_dai) {
+		for_each_rtd_cpu_dai(rtd, j, cpu_dai) {
+			dapm_add_valid_dai_widget(card, rtd,
+						  codec_dai, cpu_dai);
```
this seems odd. You will not have a connection between each CPU dai and each codec_dai

e,g. for Soundwire, you will connect ALH1.3 to RT1308.1.aif1 and ALH2.3 to RT1308.2.aif1

There will be no routing/link between e.g. ALH1.3 and RT1308.2.aif1

bardliao replied:
> this seems odd. You will not have a connection between each CPU dai and each codec_dai
> 
> e,g. for Soundwire, you will connect ALH1.3 to RT1308.1.aif1 and ALH2.3 to RT1308.2.aif1
> 
> There will be no routing/link between e.g. ALH1.3 and RT1308.2.aif1

Yes, it is what we are missing now. Thanks @plbossart

plbossart commented:
@morimoto if you have spare cycles your feedback on this PR would be more than appreciated.

bardliao commented:
@plbossart I compared and merged @morimoto 's version. Could you review it? I have two question.
1. Do we need to support multi-cpu for FE? What is the scenario of multi-cpu on codec to codec dai link?

lyakh commented on sound/soc/soc-pcm.c:
```diff
@@ -354,22 +401,40 @@ static void soc_pcm_init_runtime_hw(struct snd_pcm_substream *substream)
+		cpu_rate_max = min_not_zero(cpu_rate_max, cpu_stream->rate_max);
+		formats &= cpu_stream->formats;
+		cpu_rates = snd_pcm_rate_mask_intersect(cpu_stream->rates,
+							rates);
```
cpu_rates?

bardliao replied:
Good catch. Thanks @lyakh

lyakh commented on sound/soc/soc-dapm.c:
```diff
@@ -4295,63 +4294,83 @@ static void dapm_connect_dai_link_widgets(struct snd_soc_card *card,
+				dapm_add_valid_dai_widget(card, rtd,
+							  codec_dai,
+							  rtd->cpu_dais[i]);
+			}
```
superfluous parentheses

bardliao replied:
Updated. Thanks.

lyakh commented on sound/soc/soc-dapm.c:
```diff
@@ -4295,63 +4294,83 @@ static void dapm_connect_dai_link_widgets(struct snd_soc_card *card,
+		for_each_rtd_codec_dai(rtd, i, codec_dai) {
+			dapm_add_valid_dai_widget(card, rtd,
+						  codec_dai, rtd->cpu_dais[0]);
 		}
```
ditto

bardliao replied:
Updated. Thanks.

lyakh commented on sound/soc/soc-generic-dmaengine-pcm.c:
```diff
@@ -62,6 +62,10 @@ int snd_dmaengine_pcm_prepare_slave_config(struct snd_pcm_substream *substream,
 
+	if (rtd->num_cpus > 1)
+		dev_warn(rtd->dev,
+			 "%s doesn't support Multi CPU yet\n", __func__);
```
What would actually happen in such a case? Would it still work at least with just one CPU codec? Shouldn't we return an error in this and similar cases instrad?

bardliao replied:
> What would actually happen in such a case? Would it still work at least with just one CPU codec? Shouldn't we return an error in this and similar cases instrad?

To be honest, I don't know how to test it. I should still work with one CPU. I don't want to return an error for now is because I want to know what will happen if these functions are invoked with multi cpus.

plbossart commented:
@bardliao I've lost track of the directions here, Mark Brown seemed to think this was a good patchset, are you going to update the PR, send the changes directly to alsa-devel? Let me know, thanks!

bardliao commented:
> @bardliao I've lost track of the directions here, Mark Brown seemed to think this was a good patchset, are you going to update the PR, send the changes directly to alsa-devel? Let me know, thanks!

@plbossart Yes, I will double check the comments and send a new version to alsa-devel after addressing all comments.

dbaluta commented on sound/soc/soc-core.c:
```diff
@@ -949,44 +966,38 @@ static int soc_dai_link_sanity_check(struct snd_soc_card *card,
+				link->name);
+			return -EINVAL;
+		}
 
```
Something is off here. Either the error message is wrong or the if condition is wrong. If none of cpu->name / cpu->of_node are set we won't hit the error message. 

This is not a blocker for this PR as the issue was pre-existent. Will send an email to alsa-devel.

plbossart commented:
replaced by PR #1802
